lib/blacs.fpp

```diff
@@ -163,6 +164,23 @@ module blacs_module
 
   #:enddef blacs_gsum2d_interface
 
+
+  #:def blacs_gemr2d_interface(TYPE,PREFIX)
+
+    !> Performs 2d copy from data in matrix to another, potentially with different distribution
+    !> patterns
+    !! \see BLACS documentation for details.
+    subroutine p${PREFIX}$gemr2d(mm, nn, aa, ia, ja, descA, bb, ib, jb, descB, ictxt)
+      import
+      integer, intent(in) :: mm, nn, ia, ja, ib, jb
+      integer, intent(in) :: descA(DLEN_), descB(DLEN_)
+      ${TYPE}$, intent(in) :: aa
+      ${TYPE}$, intent(inout) :: bb
```

Comment:
```suggestion
      ${TYPE}$, intent(in) :: aa(*)
      ${TYPE}$, intent(inout) :: bb(*)
```
The two entities should be assumed sized arrays, IMO. Fortran allows to associate an array element with an assumed size array, when calling such a routine.

Reply:
The problem would be that these are 2D arrays, so LDA would also need to be passed in (i.e. aa(lda,*) ) or am I missing something obvious?

Reply:
Have changed to the suggested structure, but then have to send in (1:1,1) or similar from blacsfx.fpp

Reply:
No, you don't (and you probably also shouldn't!). Fortran actually allows shape disagreement for assumed size dummy arguments:
```
module testmod
  implicit none

contains

  subroutine assumed_size(array)
    integer, intent(inout) :: array(*)

    print *, "First 2 elements of array:", array(1), array(2)

  end subroutine assumed_size

end module testmod


program testprog
  use testmod
  implicit none

  integer :: array(2, 2)

  array(:,:) = reshape([1, 2, 3, 4], [2, 2])
  call assumed_size(array(1, 1))

end program testprog
```
The good thing on assumed size (in contrast to passing a scalar), that the standard enforces, that no copy is made.

Reply:
Doesn't seem to work for a defined interface block (instead of your example with an explicit subroutine). The below fails to build with gcc11 for interface reasons, while a call assumed_size(array(1:1, 1)) will get as far as noticing the routine for the interface is actually missing in this example:

`
module testmod
  implicit none

  interface assumed_size
    subroutine assumed_size(array)
      integer, intent(inout) :: array(*)
    end subroutine assumed_size
  end interface assumed_size

end module testmod


program testprog
  use testmod
  implicit none

  integer :: array(2, 2)

  array(:,:) = reshape([1, 2, 3, 4], [2, 2])
  call assumed_size(array(1, 1))

end program testprog
`

Reply:
OK, I've looked up the Fortran 2018 standard. Indeed, shape mismatch is only allowed in specific subroutine calls, not in generic ones. :disappointed: So, we can not do it how I've suggested.
However, if you pass a section (basically a pointer), there is no warranty, that you pass the address of the beginning of that section, and not an address to a temporary copy the compiler has created. Actually, if the section you pass is non-contiguous, the compiler is forced to create a copy. I am not sure, what would be a waterproof solution, I'll think a bit more about it. I've also posted the question to [FD](https://fortran-lang.discourse.group/t/generic-call-with-rank-mismatch/2759), maybe we get some input there.

Reply:
Couldn't we just not expose a generic interface, and instead make each specific case pubic from blacs.fypp, but then ? That name space then isn't exposed outside of blacsfx.fpp, which would put a generic interface on top for external use.

Reply:
Yes, indeed, I think that would be the best solution. Having specific declarations for each BLACS-function and call them in the type-specific wrappers.